### ISSUE TITLE
🚀 Feature Request:RPC Support for Durable Objects in `wrangler dev`

### ISSUE THREAD

**Issue by yawnt** [cloudflare/workers-sdk#5918]
### Describe the solution

When binding a durable object using RPC that was defined in a different worker, it fails with:

```
Cannot access `MyDurableObject#myMethod` as Durable Object RPC is not yet supported between multiple `wrangler dev` sessions.
```

This is really useful when using durable objects in combination with `wrangler pages dev` since, as far as I understand, you cannot deploy a worker from a Pages project.

**Comment by petebacondarwin** [comment]
Can you try extending the DurableObject class in your code, which may resolve this problem. JS RPC entry points must derive from either `DurableObject` or `WorkerEntrypoint`.

**Comment by robbertvanginkel** [comment]
Extending `DurableObject` does not fix this. The current behaviour as reported here is part of the test suite on current main, which indicates this would only be supported in the future: https://github.com/cloudflare/workers-sdk/pull/5215/commits/e469fb91a95b035642ca6b176c1f53bb15491081

**Comment by petebacondarwin** [comment]
OK thanks for the clarification. I wasn't sure if it had been fixed already. I'll add it to the backlog.

**Comment by Chinoman10** [comment]
I can also confirm this... been struggling for over a week now, causing project delays because of this...
And even if I deploy I'm unable to test this since I can't bind the DO to the Pages project (_it's the only DO that doesn't appear in the bindings list, despite being deployed like all others_).

**Comment by alexturpin** [comment]
Hey folks just chiming in with more support for this! I'm already using cross-worker RPC calls in dev, was surprised to see it didn't work on Durable Objects! Everything else is very lovely to work with, this would be the cherry on top!

**Comment by LilaRest** [comment]
Hey @alexturpin! Any updates or ETA for that feature?

**Comment by alexturpin** [comment]
@LilaRest sorry for the confusion, I'm not a Cloudflare employee, was just saying I would also find this very useful!

**Comment by LilaRest** [comment]
@alexturpin My bad, I'm quite tired and realize that I've sent the message in the wrong ticket mention the wrong person ^^ Thanks anyway for taking time to answer.

**Comment by zhihengGet** [comment]
Any update guys ? this is bad DX :(

**Comment by sassanh** [comment]
Currently, I have to deploy the worker (fast) and my next-on-pages project (slow) to test the even very small changes as long as the change has anything to do with the connection of these two and it slows the development process by an order of magnitude.

If this is going to be like this for a long time, I need to do something about it, so it would be nice if someone could provide an ETA.

A workaround would be also appreciated. If there is a hidden experimental implementation, etc., please let us know. Dealing with an unstable solution would be preferred than waiting 3-4 minutes to test a small iteration.

**Comment by grokwich** [comment]
+1

**Comment by gregory** [comment]
> OK thanks for the clarification. I wasn't sure if it had been fixed already. I'll add it to the backlog.

Hello! Could you share an update on this please?  :)
I just landed on this issue.

**Comment by petebacondarwin** [comment]
@gregory - see https://github.com/cloudflare/workers-sdk/pull/7098

**Comment by akazwz** [comment]
any news ?

**Comment by Chinoman10** [comment]
> any news ?

According to #7251 it was fixed (fix was merged) already... But I haven't tried it.

**Comment by emily-shen** [comment]
If you are not trying to communicate between Pages and Workers, #7251 should resolve this issue for you if you run all workers inside one wrangler dev session (if anyone finds any issues please let us know!).

Between Pages and Workers, we're looking into it but don't have any eta unfortunately :(
A solution for Workers + Workers Assets will probably be incoming sooner as that is a simpler problem - worked on in #7290

**Comment by zhihengGet** [comment]
as an workaround, you all can use this closed PR, it does not seem to work witch fetch tho , but hey RPC is working for me. https://github.com/cloudflare/workers-sdk/pull/7098

**Comment by bitnom** [comment]
bumping wrangler to 4.7.2 resolved this for me

**Comment by kylecarbs** [comment]
Am I confused, or should this not be closed? 

I get this when running Wrangler v4.14.1 (latest as of writing):
````
✘ [ERROR] Error: Cannot access `StreamText#potato` as Durable Object RPC is not yet supported between multiple `wrangler dev` sessions.
```

**Comment by petebacondarwin** [comment]
Yes! I think this was the point of closing this issue. If you use the single process approach to multi-workers then this does indeed work.

**Comment by kylecarbs** [comment]
@petebacondarwin is that possible using next-on-pages?

**Comment by petebacondarwin** [comment]
Since you can only connect to externally hosted Durable Objects from a Pages project there is no way to get this to work with next-on-pages. The only thing you could do is "fetch" out to a Worker from Pages and for that Worker to then bind to a Durable Object itself using the approach above.

**Comment by petebacondarwin** [comment]
That being said, we are encouraging people to consider moving away from Pages to Workers, and we have new OpenNext based Next.js tooling that lives on Workers rather than Pages. If you migrated to that, then you would get the benefits of the approach above.

**Comment by kylecarbs** [comment]
I'm doing the fetch right now! I'll look at migrating, thank you :)

**Comment by sassanh** [comment]
> If you migrated to that, then you would get the benefits of the approach above.

@petebacondarwin can you kindly elaborate on this? `initOpenNextCloudflareForDev` takes a single optional config file, how can I run multiple workers in the same process?

**Comment by mgandolfi7** [comment]
> OK thanks for the clarification. I wasn't sure if it had been fixed already. I'll add it to the backlog.

Hi @petebacondarwin, I was just wondering if this is still a planned fix/implementation? If not I'd like to request it be re-opened. I understand that the recommended way is to start all workers in a single instance, however, this is not ideal for large monorepos with multiple workers calling the same DO. 

Furthermore, even if you do chain each wrangler.jsonc in `one wrangler dev` session you lose out on HMR if you have multiple frontends bound to the same DO. Thanks.

**Comment by josephbuchma** [comment]
⚠️ You may see this error even with latest wrangler (4.46.0 at the moment) if you set `script_name` in the DO binding config. Even if you run a single `wrangler dev` instance. Remove the `script_name` to get rid of this:

> Cannot access `MyDurableObject#myMethod` as Durable Object RPC is not yet supported between multiple `wrangler dev` sessions.

---

_Took me a while to figure this out, error message does not give any clues at all. The only reason I even tried to remove the script_name was this paragraph from documentation:_

> [wrangler dev](https://developers.cloudflare.com/workers/wrangler/commands/#dev) has read access to Durable Object storage, but writes will be kept in memory and will not affect persistent data. However, if you specify the script_name explicitly in the [Durable Object binding](https://developers.cloudflare.com/workers/runtime-apis/bindings/), then writes will affect persistent data. Wrangler will emit a warning in that case.

**Comment by tstewart-klaudhaus** [comment]
I can confirm the workaround using `wrangler dev -c ./wrangler.jsonc -c ./wrangler-backend.jsonc` works, but I still get the error message when running with the cloudflare vite plugin. I have tried with various configuration combinations of `services` in wrangler.jsonc and `auxiliaryWorkers` in vite.config.ts.

Using `wrangler dev` kills productivity with SvelteKit, it would be great to see this work with the vite plugin, as suggested at:

https://developers.cloudflare.com/workers/development-testing/multi-workers/

**Comment by tstewart-klaudhaus** [comment]
There's even a test for RPC in vite dev:

https://github.com/cloudflare/workers-sdk/blob/faf8d275d04e75e88ec90328b440e0f83c0fcdba/fixtures/dev-registry/tests/dev-registry.test.ts#L559

However it never calls into the durable object rpc code branch here:

https://github.com/cloudflare/workers-sdk/blob/faf8d275d04e75e88ec90328b440e0f83c0fcdba/fixtures/dev-registry/workers/worker-entrypoint.ts#L66

Instead it passes the service name "worker-entrypoint".

The wrangler dev test DOES do durable object RPC, here:

https://github.com/cloudflare/workers-sdk/blob/faf8d275d04e75e88ec90328b440e0f83c0fcdba/fixtures/dev-registry/tests/dev-registry.test.ts#L364


 If this was enabled for vite dev as well as wrangler dev, durable object RPC would be usable for frameworks like SvelteKit.

**Comment by tstewart-klaudhaus** [comment]
Created a new issue for the Vite Dev aspect:

https://github.com/cloudflare/workers-sdk/issues/11829
